File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys

n, c = map(int, sys.stdin.readline().split())
house = []

for _ in range(n):
    house.append(int(sys.stdin.readline()))

house.sort()


start = 1
end = house[-1] - house[0]

while start <= end:
    mid = (start + end) // 2
    current = house[0]
    count = 1

    for i in house:
        if i >= current + mid:
            count += 1
            current = i

    if count >= c:
        start = mid + 1
        answer = mid
    else:
        end = mid - 1

print(answer)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.util.*;
import java.io.*;


public class Main {
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());

        arr = new int[n];
        for(int i = 0; i < n; i++){
            arr[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(arr);

        int start = 1;
        int end = arr[n-1];
        int result = 0;

        while(start <= end){
            int mid = (start + end) /2;
            int current = arr[0];
            int cnt = 1;

            for(int i = 0; i < n; i++){
                if(arr[i] >= current + mid){
                    cnt++;
                    current = arr[i];
                }
            }
            if(cnt >= c){
                result = mid;
                start = mid + 1;
            }else{
                end = mid - 1;
            }
        }
        System.out.println(result);
    }
}
